pyadjoint/optimization/tao_solver.py

```diff
@@ -0,0 +1,491 @@
+    """Interface for conversion between :class:`OverloadedType` objects and
+    :class:`petsc4py.PETSc.Vec` objects.
+
+    This uses the generic interface provided by :class:`OverloadedType`.
+
+    Args:
+        X (OverloadedType or Sequence[OverloadedType]): Defines the data
+            layout.
+        comm (petsc4py.PETSc.Comm or mpi4py.MPI.Comm): Communicator.
+    """
+
+    def __init__(self, X, *, comm=None):
+        if PETSc is None:
+            raise RuntimeError("PETSc not available")
+
+        X = Enlist(X)
+        if comm is None:
+            comm = PETSc.COMM_WORLD
+        if hasattr(comm, "tompi4py"):
+            comm = comm.tompi4py()
+
+        vecs = tuple(x._ad_to_petsc() for x in X)
+        n = sum(vec.getLocalSize() for vec in vecs)
+        N = sum(vec.getSize() for vec in vecs)
+        _, isets = PETSc.Vec().concatenate(vecs)
+        for vec in vecs:
+            vec.destroy()
+        del vecs
+
+        self._comm = comm
+        self._n = n
+        self._N = N
+        self._isets = isets
+
+        attach_destroy_finalizer(self, *self._isets)
+
+    @property
+    def comm(self):
+        """Communicator.
+        """
+
+        return self._comm
+
+    @property
+    def n(self):
+        """Total number of process local DoFs."""
+
+        return self._n
+
+    @property
+    def N(self):
+        """Total number of global DoFs."""
+
+        return self._N
+
+    def new_petsc(self):
+        """Construct a new :class:`petsc4py.PETSc.Vec`.
+
+        Returns:
+            petsc4py.PETSc.Vec: The new :class:`petsc4py.PETSc.Vec`.
+        """
+
+        vec = PETSc.Vec().create(comm=self.comm)
+        vec.setSizes((self.n, self.N))
+        vec.setUp()
+        return vec
+
+    def from_petsc(self, y, X):
+        """Copy data from a :class:`petsc4py.PETSc.Vec` to variables.
+
+        Args:
+            y (petsc4py.PETSc.Vec): The input :class:`petsc4py.PETSc.Vec`.
+            X (OverloadedType or Sequence[OverloadedType]): The output
+                variables.
+        """
+
+        X = Enlist(X)
+        if len(X) != len(self._isets):
+            raise ValueError("Invalid length")
+        for iset, x in zip(self._isets, X):
+            y_sub = y.getSubVector(iset)
+            x._ad_from_petsc(y_sub)
+            y.restoreSubVector(iset, y_sub)
+
+    def to_petsc(self, x, Y):
+        """Copy data to a :class:`petsc4py.PETSc.Vec`.
+
+        Args:
+            x (petsc4py.PETSc.Vec): The output :class:`petsc4py.PETSc.Vec`.
+            Y (Complex, OverloadedType or Sequence[OverloadedType]):
+                Values for input variables.
+        """
+
+        Y = Enlist(Y)
+        for iset, y in zip(self._isets, Y):
+            x_sub = x.getSubVector(iset)
+            if isinstance(y, Complex):
+                x_sub.set(y)
+            elif isinstance(y, OverloadedType):
+                y_vec = y._ad_to_petsc()
+                y_vec.copy(result=x_sub)
+                y_vec.destroy()
```

Comment:
```suggestion
                x_sub = y._ad_to_petsc(vec=x_sub)
```
Does the `_ad_to_petsc` with optional `vec` work better here?